File: Aether Mobile SDK/SemanticContext.kt
// =============================================================================
// AETHER SDK — Android Tiered Semantic Context
// Enriches every event with layered context based on consent + configuration.
// Tier 1: Essential (always) → Tier 2: Functional → Tier 3: Rich
// =============================================================================

package com.aether.sdk

import android.content.Context
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

// =============================================================================
// TYPES
// =============================================================================

enum class ContextTier(val level: Int) {
    ESSENTIAL(1),
    FUNCTIONAL(2),
    RICH(3);

    companion object {
        fun fromLevel(level: Int): ContextTier = entries.firstOrNull { it.level == level } ?: ESSENTIAL
    }
}

// =============================================================================
// COLLECTOR
// =============================================================================

object SemanticContextCollector {
    private const val SDK_VERSION = "5.0.0"
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    private var appContext: Context? = null

    // Tier 2 state
    private val screenPath = CopyOnWriteArrayList<String>()
    @Volatile private var eventSequence = 0
    @Volatile private var sessionStartMs = System.currentTimeMillis()
    @Volatile private var entryPoint = ""
    @Volatile private var screenDepth = 0
    @Volatile private var appState = "active"

    // Tier 3 state
    private val errorBuffer = CopyOnWriteArrayList<Triple<String, String, String>>()
    @Volatile private var backtrackCount = 0
    @Volatile private var interactionCount = 0

    // Consent
    @Volatile private var analyticsConsent = false
    @Volatile private var marketingConsent = false

    fun initialize(context: Context) {
        appContext = context.applicationContext
    }

    // =========================================================================
    // PUBLIC API
    // =========================================================================

    fun collect(): JSONObject {
        eventSequence++
        val tier = resolveTier()

        val envelope = JSONObject().apply {
            put("tier", tier.level)
            put("t1", collectTier1())
            if (tier.level >= 2) put("t2", collectTier2())
            if (tier.level >= 3) put("t3", collectTier3())
        }
        return envelope
    }

    fun recordScreen(name: String) {
        if (screenPath.size >= 2 && screenPath[screenPath.size - 2] == name) {
            backtrackCount++
        }
        screenPath.add(name)
        if (screenPath.size > 50) {
            val trimmed = screenPath.subList(screenPath.size - 50, screenPath.size).toList()
            screenPath.clear()
            screenPath.addAll(trimmed)
        }
        screenDepth++
        if (entryPoint.isEmpty()) entryPoint = name
    }

    fun recordError(message: String, type: String) {
        errorBuffer.add(Triple(message, type, dateFormat.format(Date())))
        if (errorBuffer.size > 100) errorBuffer.removeAt(0)
    }

    fun recordInteraction() { interactionCount++ }

    fun updateConsent(analytics: Boolean, marketing: Boolean) {
        analyticsConsent = analytics
        marketingConsent = marketing
    }

    fun updateAppState(state: String) { appState = state }

    fun resetSession() {
        screenPath.clear()
        eventSequence = 0
        sessionStartMs = System.currentTimeMillis()
        entryPoint = ""
        screenDepth = 0
        backtrackCount = 0
        interactionCount = 0
        errorBuffer.clear()
    }

    // =========================================================================
    // TIER COLLECTORS
    // =========================================================================

    private fun collectTier1(): JSONObject {
        val ctx = appContext
        return JSONObject().apply {
            put("eventId", UUID.randomUUID().toString())
            put("timestamp", dateFormat.format(Date()))
            put("sdkVersion", SDK_VERSION)
            put("platform", "android")
            put("device", JSONObject().apply {
                val isTablet = ctx?.resources?.configuration?.smallestScreenWidthDp?.let { it >= 600 } ?: false
                put("type", if (isTablet) "tablet" else "mobile")
                put("os", "Android ${Build.VERSION.RELEASE}")
                put("language", Locale.getDefault().language)
                put("online", true)
            })
        }
    }

    private fun collectTier2(): JSONObject {
        val elapsedMs = System.currentTimeMillis() - sessionStartMs
        return JSONObject().apply {
            put("journeyStage", inferJourneyStage(elapsedMs))
            put("screenPath", JSONArray(screenPath.takeLast(20)))
            put("sessionDuration", elapsedMs)
            put("appState", appState)
            put("screenDepth", screenDepth)
            put("eventSequenceIndex", eventSequence)
            put("entryPoint", entryPoint.ifEmpty { screenPath.firstOrNull() ?: "unknown" })
        }
    }

    private fun collectTier3(): JSONObject {
        val windowMs = 300_000L
        val cutoff = System.currentTimeMillis() - windowMs
        val recentErrors = errorBuffer.filter {
            try { dateFormat.parse(it.third)?.time?.let { ts -> ts >= cutoff } ?: false } catch (_: Exception) { false }
        }
        val errorRate = recentErrors.size.toDouble() / (windowMs / 60_000.0)

        val elapsedS = (System.currentTimeMillis() - sessionStartMs) / 1000.0
        val engagement = minOf(1.0, (screenDepth / 10.0) * 0.4 + minOf(elapsedS / 120.0, 1.0) * 0.6)
        val frustration = minOf(1.0, recentErrors.size * 0.2 / 5.0)
        val confusion = minOf(1.0, backtrackCount * 0.4 / 3.0)
        val urgency = if (eventSequence > 0) minOf(1.0, (screenDepth.toDouble() / eventSequence) * 2.0) else 0.0

        val lastErr = errorBuffer.lastOrNull()

        return JSONObject().apply {
            put("inferredIntent", JSONObject.NULL)
            put("sentimentSignals", JSONObject().apply {
                put("frustration", (frustration * 1000).toLong() / 1000.0)
                put("engagement", (engagement * 1000).toLong() / 1000.0)
                put("urgency", (urgency * 1000).toLong() / 1000.0)
                put("confusion", (confusion * 1000).toLong() / 1000.0)
            })
            put("errorLog", JSONObject().apply {
                put("errorCount", errorBuffer.size)
                if (lastErr != null) {
                    put("lastError", JSONObject().apply {
                        put("message", lastErr.first)
                        put("type", lastErr.second)
                        put("timestamp", lastErr.third)
                    })
                } else {
                    put("lastError", JSONObject.NULL)
                }
                put("errorRate", (errorRate * 100).toLong() / 100.0)
            })
        }
    }

    // =========================================================================
    // HELPERS
    // =========================================================================

    private fun resolveTier(): ContextTier {
        if (analyticsConsent && marketingConsent) return ContextTier.RICH
        if (analyticsConsent) return ContextTier.FUNCTIONAL
        return ContextTier.ESSENTIAL
    }

    private fun inferJourneyStage(elapsedMs: Long): String {
        if (elapsedMs < 10_000 && screenDepth <= 1) return "new"
        if (screenDepth > 5 || elapsedMs > 180_000) return "engaged"
        return "returning"
    }
}

File: Aether Mobile SDK/SemanticContext.swift
// =============================================================================
// Aether SDK — iOS Tiered Semantic Context
// Enriches every event with layered context based on consent + configuration.
// Tier 1: Essential (always) → Tier 2: Functional → Tier 3: Rich
// =============================================================================

import Foundation
import UIKit

// MARK: - Types

public enum ContextTier: Int, Codable, Comparable {
    case essential = 1
    case functional = 2
    case rich = 3

    public static func < (lhs: ContextTier, rhs: ContextTier) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

public struct Tier1Context: Codable {
    public let eventId: String
    public let timestamp: String
    public let sdkVersion: String
    public let platform: String
    public let device: Tier1Device

    public struct Tier1Device: Codable {
        public let type: String
        public let os: String
        public let language: String
        public let online: Bool
    }
}

public struct Tier2Context: Codable {
    public let journeyStage: String
    public let screenPath: [String]
    public let sessionDuration: Double
    public let appState: String
    public let screenDepth: Int
    public let eventSequenceIndex: Int
    public let entryPoint: String
}

public struct Tier3Context: Codable {
    public let inferredIntent: InferredIntent?
    public let sentimentSignals: SentimentSignals
    public let errorLog: ErrorLog

    public struct InferredIntent: Codable {
        public let action: String
        public let confidence: Double
        public let journeyPhase: String
    }

    public struct SentimentSignals: Codable {
        public let frustration: Double
        public let engagement: Double
        public let urgency: Double
        public let confusion: Double
    }

    public struct ErrorLog: Codable {
        public let errorCount: Int
        public let lastError: ErrorEntry?
        public let errorRate: Double

        public struct ErrorEntry: Codable {
            public let message: String
            public let type: String
            public let timestamp: String
        }
    }
}

public struct SemanticContextEnvelope: Codable {
    public let tier: Int
    public let t1: Tier1Context
    public let t2: Tier2Context?
    public let t3: Tier3Context?
}

// MARK: - Collector

public final class SemanticContextCollector {
    public static let shared = SemanticContextCollector()

    private let sdkVersion = "5.0.0"
    private let serialQueue = DispatchQueue(label: "com.aether.sdk.context")

    // Tier 2 state
    private var screenPath: [String] = []
    private var eventSequence: Int = 0
    private var sessionStartDate: Date = Date()
    private var entryPoint: String = ""
    private var screenDepth: Int = 0
    private var appState: String = "active"

    // Tier 3 state
    private var errorBuffer: [(message: String, type: String, timestamp: String)] = []
    private var backtrackCount: Int = 0
    private var sessionInteractionCount: Int = 0
    private var lastIntent: Tier3Context.InferredIntent? = nil

    // Consent
    private var analyticsConsent: Bool = false
    private var marketingConsent: Bool = false

    private init() {
        setupLifecycleObservers()
    }

    // MARK: - Public API

    public func collect() -> SemanticContextEnvelope {
        return serialQueue.sync {
            eventSequence += 1
            let tier = resolveTier()

            let t1 = collectTier1()
            let t2: Tier2Context? = tier >= .functional ? collectTier2() : nil
            let t3: Tier3Context? = tier >= .rich ? collectTier3() : nil

            return SemanticContextEnvelope(tier: tier.rawValue, t1: t1, t2: t2, t3: t3)
        }
    }

    public func recordScreen(_ name: String) {
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            if self.screenPath.count >= 2 && self.screenPath[self.screenPath.count - 2] == name {
                self.backtrackCount += 1
            }
            self.screenPath.append(name)
            if self.screenPath.count > 50 { self.screenPath = Array(self.screenPath.suffix(50)) }
            self.screenDepth += 1
        }
    }

    public func recordError(message: String, type: String) {
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            self.errorBuffer.append((message: message, type: type, timestamp: ISO8601DateFormatter().string(from: Date())))
            if self.errorBuffer.count > 100 { self.errorBuffer.removeFirst() }
        }
    }

    public func recordInteraction() {
        serialQueue.async { [weak self] in
            self?.sessionInteractionCount += 1
        }
    }

    public func updateConsent(analytics: Bool, marketing: Bool) {
        serialQueue.async { [weak self] in
            self?.analyticsConsent = analytics
            self?.marketingConsent = marketing
        }
    }

    public func resetSession() {
        serialQueue.async { [weak self] in
            self?.screenPath = []
            self?.eventSequence = 0
            self?.sessionStartDate = Date()
            self?.screenDepth = 0
            self?.backtrackCount = 0
            self?.sessionInteractionCount = 0
            self?.errorBuffer = []
        }
    }

    // MARK: - Tier Collectors

    private func collectTier1() -> Tier1Context {
        let device = UIDevice.current
        return Tier1Context(
            eventId: UUID().uuidString,
            timestamp: ISO8601DateFormatter().string(from: Date()),
            sdkVersion: sdkVersion,
            platform: "ios",
            device: .init(
                type: device.userInterfaceIdiom == .pad ? "tablet" : "mobile",
                os: "iOS \(device.systemVersion)",
                language: Locale.current.language.languageCode?.identifier ?? "en",
                online: true // Reachability would be injected in production
            )
        )
    }

    private func collectTier2() -> Tier2Context {
        let elapsed = Date().timeIntervalSince(sessionStartDate)
        return Tier2Context(
            journeyStage: inferJourneyStage(elapsed: elapsed),
            screenPath: Array(screenPath.suffix(20)),
            sessionDuration: elapsed * 1000,
            appState: appState,
            screenDepth: screenDepth,
            eventSequenceIndex: eventSequence,
            entryPoint: entryPoint.isEmpty ? (screenPath.first ?? "unknown") : entryPoint
        )
    }

    private func collectTier3() -> Tier3Context {
        let windowSeconds: Double = 300
        let recentErrors = errorBuffer.filter {
            guard let ts = ISO8601DateFormatter().date(from: $0.timestamp) else { return false }
            return Date().timeIntervalSince(ts) < windowSeconds
        }
        let errorRate = Double(recentErrors.count) / (windowSeconds / 60.0)

        let elapsed = Date().timeIntervalSince(sessionStartDate)
        let engagement = min(1.0, (Double(screenDepth) / 10.0) * 0.4 + min(elapsed / 120.0, 1.0) * 0.6)
        let frustration = min(1.0, Double(recentErrors.count) * 0.2 / 5.0)
        let confusion = min(1.0, Double(backtrackCount) * 0.4 / 3.0)
        let urgency = min(1.0, eventSequence > 0 ? (Double(screenDepth) / Double(eventSequence)) * 2.0 : 0)

        let lastErr: Tier3Context.ErrorLog.ErrorEntry? = errorBuffer.last.map {
            .init(message: $0.message, type: $0.type, timestamp: $0.timestamp)
        }

        return Tier3Context(
            inferredIntent: lastIntent,
            sentimentSignals: .init(
                frustration: (frustration * 1000).rounded() / 1000,
                engagement: (engagement * 1000).rounded() / 1000,
                urgency: (urgency * 1000).rounded() / 1000,
                confusion: (confusion * 1000).rounded() / 1000
            ),
            errorLog: .init(
                errorCount: errorBuffer.count,
                lastError: lastErr,
                errorRate: (errorRate * 100).rounded() / 100
            )
        )
    }

    // MARK: - Helpers

    private func resolveTier() -> ContextTier {
        if analyticsConsent && marketingConsent { return .rich }
        if analyticsConsent { return .functional }
        return .essential
    }

    private func inferJourneyStage(elapsed: TimeInterval) -> String {
        if elapsed < 10 && screenDepth <= 1 { return "new" }
        if screenDepth > 5 || elapsed > 180 { return "engaged" }
        return "returning"
    }

    private func setupLifecycleObservers() {
        NotificationCenter.default.addObserver(forName: UIApplication.didBecomeActiveNotification, object: nil, queue: .main) { [weak self] _ in
            self?.serialQueue.async { self?.appState = "active" }
        }
        NotificationCenter.default.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main) { [weak self] _ in
            self?.serialQueue.async { self?.appState = "background" }
        }
        NotificationCenter.default.addObserver(forName: UIApplication.willResignActiveNotification, object: nil, queue: .main) { [weak self] _ in
            self?.serialQueue.async { self?.appState = "inactive" }
        }
    }
}
